RFC: Metatable type functions (#69)
luau-lang/rfcs: docs/metatable-type-functions.md

# Metatable Type Functions

## Summary

Implement type functions for `getmetatable` and `setmetatable`.

## Motivation

### `setmetatable` Type Function

There is currently no way for users to apply metatable type information to a type without the usage of `typeof()`. This isn't ideal, as it adds verbosity and boilerplate to common patterns such as object-oriented programming. For example, the following:

> ```lua

Reviewer: ```suggestion
```luau
```

local clock = {}
type Identity = typeof(setmetatable({} :: { time: number }, { __index = clock }))
```

could be reduced to:

> ```lua

Reviewer: ```suggestion
```luau
```

local clock = {}

> type Identity = setmetatable<{ time: number }, { __index: clock }>

Reviewer: Did you mean to use `typeof(clock)` here?

```

### `getmetatable` Type Function

[Issue #1435](https://github.com/luau-lang/luau/issues/1435) is caused by a lack of special behavior for `getmetatable` when the type's metatable has a `__metatable` field. This is fixable in a variety of ways, however a type function has multiple benefits and exposes more type-level expression for users. It fits in nicely with the new solver, as type functions are a cleaner alternative to magic functions.

## Design

> ### `setmetatable` Type Function

Reviewer: What is the semantics of `setmetatable` when the first argument is not a table (or union/intersection of tables)? Presumably, reduction failure? This should be specified.

In the following code example, `Identity` should evaluate to `{ sound: string, @metatable: { __index: animal } }`:

> ```lua

Reviewer: ```suggestion
```luau
```

local animal = {}
type Identity = setmetatable<{
	sound: string
}, {

> 	__index: animal

Reviewer: Do you mean for this to be `typeof(animal)`? Or for `animal` to be a type?

}>
```

> ### `getmetatable` Type Function

Reviewer: What is the semantics of `getmetatable` when the argument does not have a metatable? What about when it cannot have one, e.g. because it is neither a table nor class? What does it do when given a union of tables? Should it check if the tables all have the same metatable and return the one or error otherwise? Should it return a union of the metatables across the union?

Reviewer: My general impression is that it makes the most sense to air conservatively here, we can always specify more behavior later. Possible answer: `getmetatable` will work on a table, and a class, and give you their respective metatables, but nothing else, and will respect `__metatable`.

In the following code example, `ReversedIdentity` should evaluate to `{ __index: animal }`:

> ```lua

Reviewer: ```suggestion
```luau
```

local animal = {}
type Identity = setmetatable<{
	sound: string
}, {

> 	__index: animal

Reviewer: Do you mean for this to be `typeof(animal)`? Or for `animal` to be a type?

Author: Yeah, I did. Oops

}>

type ReversedIdentity = getmetatable<Identity>
```

Due to `__metamethod`, additional behavior needs to be met. In the following code example, `MetatableType` should evaluate to `"No metatable here!"`:

> ```lua

Reviewer: ```suggestion
```luau
```

## Design

### `setmetatable` Type Function

In the following code example, `Identity` should evaluate to `{ sound: string, @metatable: { __index: typeof(animal) } }`:

```luau
local animal = {}
type Identity = setmetatable<{
	sound: string
}, {
	__index: typeof(animal)
}>
```

When a non-table type (number, boolean) is passed into the type function, it should result in an error. This includes classes.
If the type is an intersection or union of tables, then the metatable should be applied to every component type.
If the second argument of the type function isn't a table, then it should error.

### `getmetatable` Type Function

In the following code example, `ReversedIdentity` should evaluate to `{ __index: typeof(animal) }`:

```luau
local animal = {}
type Identity = setmetatable<{
	sound: string
}, {
	__index: typeof(animal)
}>

type ReversedIdentity = getmetatable<Identity>
```

> Due to `__metamethod`, additional behavior needs to be met. In the following code example, `MetatableType` should evaluate to `"No metatable here!"`:

Reviewer 2: ```suggestion
Due to `__metatable`, additional behavior needs to be met. In the following code example, `MetatableType` should evaluate to `"No metatable here!"`:
```

```luau
type metatable = {
	__metatable: "No metatable here!"
}

type MetatableType = getmetatable<setmetatable<{}, metatable>>
```

> If the type passed into the type function doesn't have a metatable and it *could* have a metatable (so primitives, singletons, classes, and tables), then it will return nil.
> If the type passed in is an intersection or union, then it will return an intersection or union respectively.

Reviewer: ```suggestion
If the type passed in does not support metatables in the type system (i.e. at the time of writing, is not any of a primitive, a singleton, a class, a table, or a union or intersection thereof), it should error when passed to `getmetatable`. 
If the type passed in is an intersection or union of things that support metatables, then it will return an intersection or union of those metatables respectively.
If no metatable is present, but the type system supports metatables for that type, it should return the `nil` type indicating that there is no metatable present.
```

# Metatable Type Functions

## Summary

Implement type functions for `getmetatable` and `setmetatable`.

## Motivation

### `setmetatable` Type Function

There is currently no way for users to apply metatable type information to a type without the usage of `typeof()`. This isn't ideal, as it adds verbosity and boilerplate to common patterns such as object-oriented programming. For example, the following:

```luau
local clock = {}
type Identity = typeof(setmetatable({} :: { time: number }, { __index = clock }))
```

could be reduced to:

```luau
local clock = {}
type Identity = setmetatable<{ time: number }, { __index: typeof(clock) }>
```

> If the second argument of the type function isn't a table, then it should error.

Reviewer: This should be in the design section.

### `getmetatable` Type Function

[Issue #1435](https://github.com/luau-lang/luau/issues/1435) is caused by a lack of special behavior for `getmetatable` when the type's metatable has a `__metatable` field. This is fixable in a variety of ways, however a type function has multiple benefits and exposes more type-level expression for users. It fits in nicely with the new solver, as type functions are a cleaner alternative to magic functions.

## Design

### `setmetatable` Type Function

In the following code example, `Identity` should evaluate to `{ sound: string, @metatable: { __index: typeof(animal) } }`:

```luau
local animal = {}
type Identity = setmetatable<{
	sound: string
}, {
	__index: typeof(animal)
}>
```

When a non-table type (number, boolean) is passed into the type function, it should result in an error. This includes classes.

> If the type is an intersection/union of tables, then the metatable should be applied to every 

Reviewer: Stopped writing midsentence?
```suggestion
If the type is an intersection or union of tables, then the metatable should be applied to every component type.
```

### `getmetatable` Type Function

In the following code example, `ReversedIdentity` should evaluate to `{ __index: typeof(animal) }`:

```luau
local animal = {}
type Identity = setmetatable<{
	sound: string
}, {
	__index: typeof(animal)
}>

type ReversedIdentity = getmetatable<Identity>
```

Due to `__metamethod`, additional behavior needs to be met. In the following code example, `MetatableType` should evaluate to `"No metatable here!"`:

```luau
type metatable = {
	__metatable: "No metatable here!"
}

type MetatableType = getmetatable<setmetatable<{}, metatable>>
```

If the type passed into the type function doesn't have a metatable and it *could* have a metatable (so primitives, singletons, classes, and tables), then it will return nil.

> If the type passed in is an intersection/union, then it will return an interesction/union respectively.

Reviewer: ```suggestion
If the type passed in is an intersection or union, then it will return an intersection or union respectively.
```
